pkg/cover/backend/mach-o.go

```diff
@@ -0,0 +1,146 @@
+// Copyright 2020 syzkaller project authors. All rights reserved.
+// Use of this source code is governed by Apache 2 LICENSE that can be found in the LICENSE file.
+
+package backend
+
+import (
+	"debug/macho"
+	"fmt"
+	"path/filepath"
+	"sort"
+	"strings"
+
+	"github.com/google/syzkaller/pkg/host"
+	"github.com/google/syzkaller/sys/targets"
+)
+
+func makeMachO(target *targets.Target, objDir, srcDir, buildDir string,
+	hostModules []host.KernelModule) (*Impl, error) {
+	// TODO: Linux/ELF supports module symbols. We should probably also do that
+	// for XNU/Mach-O. To maximize code re-use we already have a lot of the
+	// plumbing for module support. I think we mainly miss an equivalent to
+	// discoverModules and elfReadModuleCoverPoints at the moment.
+	if len(hostModules) != 0 {
+		return nil, fmt.Errorf("module coverage not yet supported for Mach-O")
+	}
+
+	return makeDWARF(target, objDir, srcDir, buildDir,
+		[]*Module{
+			// A dummy module representing the kernel itself.
+			{
+				Name: "",
+				Path: filepath.Join(objDir, target.KernelObject),
+				Addr: 0,
+			},
+		},
+		&containerFns{
+			readSymbols:           machoReadSymbols,
+			readCoverPoints:       machoReadCoverPoints,
+			readModuleCoverPoints: machoReadModuleCoverPoints,
+			readTextRanges:        machoReadTextRanges,
+		},
+	)
+}
+
+func machoReadSymbols(module *Module) ([]*Symbol, *symbolInfo, error) {
+	file, err := macho.Open(module.Path)
+	if err != nil {
+		return nil, nil, err
+	}
+	text := file.Section("__text")
+	if text == nil {
+		return nil, nil, fmt.Errorf("no __text section in the object file")
+	}
+	if file.Symtab == nil {
+		return nil, nil, fmt.Errorf("failed to read Mach-O symbols")
+	}
+	var symbols []*Symbol
+	info := &symbolInfo{
```

Comment:
This is duplicated across elf/macho... maybe we should pass in info with already created maps?

Reply:
ACK.